File: src/boj/BOJ_1781.java
package boj;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

/**
 * {@link}  https://www.acmicpc.net/problem/1781
 * @date    2020-10-17
 * @author  lolol0705
 * @memory  81520KB / 256MB
 * @time    1008ms / 2초
 */
public class BOJ_1781 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        LinkedList<int[]> list = new LinkedList<>();
        for (int i = 0; i < n; ++i) {
            StringTokenizer st = new StringTokenizer(br.readLine(), " ");
            int deadline = Integer.parseInt(st.nextToken());
            int cup = Integer.parseInt(st.nextToken());
            list.add(new int[] {deadline, cup});
        }
        list.sort((o1, o2) -> Integer.compare(o2[0], o1[0]));

        int cnt = 0;
        PriorityQueue<Integer> pq = new PriorityQueue<>((o1, o2) -> Integer.compare(o2, o1));
        for (int i = n; i >= 1; --i) {
            while (!list.isEmpty() && i <= list.peek()[0]) {
                pq.add(list.poll()[1]);
            }
            if (!pq.isEmpty()) {
                cnt += pq.poll();
            }
        }
        System.out.print(cnt);
    }
}

File: src/boj/BOJ_1781.kt
package boj

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

/**
 * {@link}  https://www.acmicpc.net/problem/1939
 * @date    2020-10-17
 * @author  lolol0705
 * @comment 백준 JDK로인해 컴파일 에러이지만 자바와 동일
 */
fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n = br.readLine().toInt()
    val list = LinkedList<IntArray>()
    repeat(n) {
        val st = StringTokenizer(br.readLine(), " ")
        val deadline = st.nextToken().toInt()
        val cup = st.nextToken().toInt()
        list.add(intArrayOf(deadline, cup))
    }
    list.sortWith { o1, o2 -> o2[0].compareTo(o1[0]) }

    var cnt = 0
    val pq = PriorityQueue<Int> { o1, o2 -> o2.compareTo(o1) }
    repeat(n) {
        val cur = n-it
        while (list.isNotEmpty() && cur <= list.first[0]) {
            pq.add(list.poll()[1])
        }
        if (pq.isNotEmpty()) {
            cnt += pq.poll()
        }
    }
    print(cnt)
}